### ISSUE TITLE
MapOf from Iterator<Map.Entry<X, Y>> fails to process all entries.

### ISSUE THREAD

**Issue by driver733** [yegor256/cactoos#659]
```
@Test
    public void createsMapFromIterator() {
        MatcherAssert.assertThat(
            "Can't create a map from Iterator with Entries.",
            new MapOf<Integer, Integer>(
                new ListOf<Map.Entry<Integer, Integer>>(
                    new MapEntry<Integer, Integer>(0, 0),
                    new MapEntry<Integer, Integer>(1, 1)
                ).iterator()
            ),
            Matchers.allOf(
                Matchers.hasEntry(0, 0),
                Matchers.hasEntry(1, 1)
            )
        );
    }
```

Expected:
`(map containing [<1>-><1>] and map containing [<0>-><0>])`
Actual:
`map containing [<1>-><1>] map was []`

Full output:
```
java.lang.AssertionError: Can't create a map from Iterator with Entries.
Expected: (map containing [<1>-><1>] and map containing [<0>-><0>])
     but: map containing [<1>-><1>] map was []

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.cactoos.map.MapOfTest.createsMapFromIterator(MapOfTest.java:166)
```

p.s. FYI, if the matchers are used separately (without `allOf`) then the test passes. I believe that some investigation is required.

**Comment by 0crat** [comment]
@yegor256[/z](https://www.0crat.com/u/yegor256) please, pay attention to this issue

**Comment by 0crat** [comment]
Job [#659](https://github.com/yegor256/cactoos/issues/659) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
Bug was reported, see [§29](http://datum.zerocracy.com/pages/policy.html#29): +15 points just awarded to @driver733[/z](https://www.0crat.com/u/driver733), total is [+150](https://www.0crat.com/u/driver733)

**Comment by 0crat** [comment]
The job [#659](https://github.com/yegor256/cactoos/issues/659) assigned to @sharpler[/z](https://www.0crat.com/u/sharpler), here is [why](https://www.0crat.com/footprint/C63314D6Z/471612081). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by SharplEr** [comment]
Wow, It's happening because you can use iterator only once.
Like that:
```Java
    Map<Integer, Integer> map = new MapOf<Integer, Integer>(
            new ListOf<Map.Entry<Integer, Integer>>(
                new MapEntry<Integer, Integer>(0, 0),
                new MapEntry<Integer, Integer>(1, 1)
            ).iterator()
        );
        System.out.println(map.size());
        System.out.println(map.size());
```
Result will be

```
2
0
```

Nice:)

**Comment by SharplEr** [comment]
@yegor256 @driver733
As I wrote in the last comment the problem is iterator has mutable state and can't be used twice.
So I see only 3 possible solutions:
1) Remove constructor with an iterator. Iterator is a bad mutable class anyway :)
2) Put in constructor all elements into new ArrayList and pass that new list into constructor with iterable. But this violates a rule about code in constructors.
3) Create lazy repeatable iterator class which takes a normal iterator and lazy saves all elements into a private mutable ArrayList and will return elements from that list when trying to use it twice. But this violates a rule about mutable.

Please remember that there are a lot of constructors with iterator in the project. So the solution what we will choose here will be applied to many classes.

**Comment by SharplEr** [comment]
It also connected with https://github.com/yegor256/cactoos/issues/625
@0crat waiting

**Comment by 0crat** [comment]
Are you speaking to me or about me [here](https://github.com/yegor256/cactoos/issues/659#issuecomment-366366520). You must always start your message with my name if you want to address it to me, see [§1](http://datum.zerocracy.com/pages/policy.html#1).

**Comment by SharplEr** [comment]
@0crat waiting

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/yegor256/cactoos/issues/659#issuecomment-366442790))

@SharplEr The impediment for [#659](https://github.com/yegor256/cactoos/issues/659) was registered successfully by @sharpler[/z](https://www.0crat.com/u/sharpler)

**Comment by SharplEr** [comment]
@yegor256 @driver733 as I understood from #677 I should remove ctor with iterator.

**Comment by yegor256** [comment]
@SharplEr I think so, yes

**Comment by SharplEr** [comment]
@driver733 I'm still waiting for any reviewer reaction.

**Comment by SharplEr** [comment]
@0crat waiting

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/yegor256/cactoos/issues/659#issuecomment-370366580))

@SharplEr Job [#659](https://github.com/yegor256/cactoos/issues/659) is already on hold

**Comment by SharplEr** [comment]
@driver733 PR #712  was merged, please close the issue.

**Comment by 0crat** [comment]
@elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova) please review this job completed by @sharpler[/z](https://www.0crat.com/u/sharpler), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#659](https://github.com/yegor256/cactoos/issues/659) is now out of scope

**Comment by elenavolokhova** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +30 points just awarded to @sharpler[/z](https://www.0crat.com/u/sharpler)

**Comment by 0crat** [comment]
Quality review completed: +8 points just awarded to @elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova)
